src/unit-of-work/index.js

```diff
@@ -0,0 +1,480 @@
+import Utils from '../utils'
+import Async from '../request/async'
+import DataQueryBuilder from '../data/query-builder'
+
+import { OperationType, IsolationLevelEnum } from './constants'
+import { OpResult } from './op-result'
+import { OpResultValueReference } from './op-result-value-reference'
+
+function resolveTableName(obj) {
+  return Utils.getClassName(obj)
+}
+
+class UnitOfWork {
+  static IsolationLevelEnum = IsolationLevelEnum
+
+  constructor(isolationLevelEnum, app) {
+    this.app = app
+
+    this.payload = {
+      isolationLevelEnum,
+      operations: []
+    }
+
+    this.usedOpIds = {}
+  }
+
+  getOpStackName(operationType, table) {
+    return `${operationType.toLowerCase()}${table}`
+  }
+
+  getNextOpResultIndex(stackName) {
+    if (!this.usedOpIds[stackName]) {
+      this.usedOpIds[stackName] = 0
+    }
+
+    return ++this.usedOpIds[stackName]
+  }
+
+  addOperations(operationType, table, payload) {
+    const opResult = new OpResult(this, { operationType, table, payload })
+
+    this.payload.operations.push({
+      operationType,
+      table,
+      payload,
+      meta: {
+        opResult
+      }
+    })
+
+    return opResult
+  }
+
+  composePayload() {
+    return {
+      ...this.payload,
+      operations: this.payload.operations.map(({ meta, ...operation }) => {
+        return {
+          ...operation,
+          opResultId: meta.opResult.getOpResultId(),
+        }
+      })
+    }
+  }
+
+  execute() {
+    return Promise.resolve()
+      .then(() => new Promise((resolve, reject) => {
+        return this.app.request.post({
+          url         : this.app.urls.transactions(),
+          data        : this.composePayload(),
+          asyncHandler: new Async(resolve, reject),
+        })
+      }))
+      .then(result => {
+        if (result.results) {
+          this.payload.operations.forEach(operation => {
+            const opResultId = operation.meta.opResult.getOpResultId()
+
+            if (result.results[opResultId]) {
+              operation.meta.opResult.setResult(result.results[opResultId].result)
+            }
+          })
+        }
+
+        return result
+      })
```

Comment:
I suggest to rework this method with async/await, maybe it will become more clear and simpler

Reply:
I agree

Reply:
unfortunately `async/await` is not supported yet, so I'm waiting for this one =) https://github.com/Backendless/JS-SDK/pull/151#discussion_r388541475

Reply:
yes, please

Reply:
Done